### ISSUE TITLE
Discussion: opt-out of double symbol lookup for certain contextual keywords

### ISSUE THREAD

**Issue by alrz** [dotnet/csharplang#817]
Currently, some contextual keywords cause a redundant symbol lookup, namely, `nameof`, `var` and `_` (and later `let` etc). Opting out of this causes a faster compile time and removes some ambiguities. Also, it would eliminate the confusion around discards vs. `_` identifier. Since most of projects don't even use these as valid identifiers per a standard code convention, there will be some room to simplify parsing and binding with this feature enabled. Analyzers can help to prevent using these as symbol names, but since the compiler wouldn't know that, it still runs a double lookup for each instance.

**Comment by Joe4evr** [comment]
![](https://emojipedia-us.s3.amazonaws.com/thumbs/160/facebook/105/thinking-face_1f914.png)
It's an interesting idea, but is it really *that much* of a factor when it comes to source->IL compilation? I mean, sure every little bit helps, but I doubt that the compilation of most codebases (even very large ones) are choking on the double symbol lookup for contextual keywords.

**Comment by HaloFour** [comment]
While it would be nice to alleviate the contextual keyword burden from the compiler and improve performance all this option would do is result in a separate dialect of C#.

**Comment by alrz** [comment]
>separate dialect of C#.

It bothered me so much I haven't given it much thought. Yes, that I think is a no-go.

**Comment by Joe4evr** [comment]
Not so fast, I think there is still something that could potentially be done here. Specifically regarding:
> [the compiler] still runs a double lookup for each instance.

The compiler could memoize whether no types/members with the same name as a contextual keyword are declared as part of one of its passes. That way, it shouldn't need to check more than once for the vast, vast majority of codebases.
```cs
if (ProjectContainsTypeOrMemberWithContextualName)
{
    TryParseAsTypeOrMember();
}
TryParseAsKeyword();
```

**Comment by alrz** [comment]
whether or not a name exists, depends on the context in which it appears wrt using directives, nesting, etc.

A simple example is parameter names: you'd need to cache e.g. `var` lookup, for each method individually. I'm don't think that it would worth it considering complexity cost and memory usage.

**Comment by yaakov-h** [comment]
@Joe4evr Wouldn't that be a pure Roslyn optimisation, not a C# language change?

**Comment by Joe4evr** [comment]
> whether or not a name exists, depends on the context in which it appears wrt using directives, nesting, etc.

Not entirely, given the [order of passes](https://blogs.msdn.microsoft.com/ericlippert/2010/02/04/how-many-passes/), the compiler knows ***all*** the type and member names before it even starts to generate some IL (assuming the gist of that post is still accurate when it comes to Roslyn).

Of course, the `ProjectContainsTypeOrMemberWithContextualName` would be virally `true` if a quick metadata scan of all dependencies reveals that one such type/member is declared in any dependency. But I seriously doubt that any legit codebase does that since it violates naming guidelines and creates ambiguity for the reader. I've only seen it done with example code to describe exactly this behavior of binding to a type/member before it's checked as a keyword.

**Comment by alrz** [comment]
@Joe4evr

>given the order of passes, the compiler knows all the type and member names before it even starts to generate some IL

What about type and members from metadata i.e. referenced assemblies?

>if a quick metadata scan of all dependencies reveals that one such type/member is declared

Who is responsible for this "quick scan"? I don't think the compiler would do that. In case you're proposing to emit a new annotation to indicate that, see below.

>the ProjectContainsTypeOrMemberWithContextualName would be virally true

Interestingly, there is that kind of annotation for extension methods,

> ```
> public override bool MightContainExtensionMethods
> {
>     get
>     {
>         // While the specification for ExtensionAttribute requires that the containing assembly
>         // have the attribute if any type in the assembly has the attribute, some compilers do
>         // not properly follow that spec. Therefore we pessimistically assume every assembly
>         // may contain extension methods.
>         return true;
>     }
> }
> ```

But as you can see, it's not practical.

> seriously doubt that any legit codebase does that since it violates naming guidelines and creates ambiguity

If it's permitted, it's possible and *will* happen. I've read in some other issue (design notes) that people intentionally use `class var {}` to forbid use of `var`. Yes, wow.

**Comment by alrz** [comment]
I'll open this if anyone else had ideas to share.

**Comment by Joe4evr** [comment]
> Who is responsible for this "quick scan"? I don't think the compiler would do that.

How else is the compiler able to give IntelliSense/Object Browser in the IDE for types in referenced assemblies?

**Comment by alrz** [comment]
>to give IntelliSense/Object Browser in the IDE

IDE?

**Comment by CyrusNajmabadi** [comment]
> Opting out of this causes a faster compile time 

How much faster?

Thanks!

**Comment by alrz** [comment]
It won't be more than the time spend on a redundant using directive, since the compiler needs to scan all imported types to flag duplicates. As I quoted above, there is a similar mechanism to prevent additional lookups if the assembly doesn not contain any extension methods but it's disabled because there is a chance that not all compilers follow that spec. So I concluded that this is not practical in this way or the other and eventually the cost outweighs the performance benefit.
